### PR TITLE
[flang][OpenMP] Parsing support for iterator modifiers in FROM and TO

### PR DESCRIPTION
**Pull request by kparzysz**
Parse PRESENT modifier as well while we're at it (no MAPPER though). Add semantic checks for these clauses in the TARGET UPDATE construct, TODO messages in lowering.

### CODE REVIEW

File: flang/lib/Parser/unparse.cpp [hunk 1]
```diff
@@ -2138,6 +2138,27 @@ class UnparseVisitor {
     Put(",");
     Walk(std::get<std::optional<ScalarIntConstantExpr>>(x.t));
   }
+  void Unparse(const OmpFromClause &x) {
+    auto &expect =
+        std::get<std::optional<std::list<OmpFromClause::Expectation>>>(x.t);
+    auto &iter = std::get<std::optional<std::list<OmpIteratorModifier>>>(x.t);
```
Comment (kiranchandramohan) [inline flang/lib/Parser/unparse.cpp:2144]: Nit: braced initialization.

File: flang/lib/Parser/openmp-parsers.cpp [hunk 2]
```diff
@@ -122,6 +122,40 @@ template <typename Separator> struct MapModifiers {
...
+  // Parsing of mappers if not implemented yet.
+  using ExpParser = Parser<OmpFromClause::Expectation>;
+  using IterParser = Parser<OmpIteratorModifier>;
+  using ModParser = ConcatSeparated<Separator, ExpParser, IterParser>;
+
+  using resultType = typename ModParser::resultType;
+
+  std::optional<resultType> Parse(ParseState &state) const {
+    auto mp = ModParser(sep_, ExpParser{}, IterParser{});
+    auto mods = mp.Parse(state);
```
Comment (kiranchandramohan) [inline flang/lib/Parser/openmp-parsers.cpp:143]: Nit: Braced initialization.

File: flang/lib/Parser/unparse.cpp [hunk 3]
```diff
@@ -2241,6 +2262,27 @@ class UnparseVisitor {
     Walk(":",
         std::get<std::optional<OmpDefaultmapClause::VariableCategory>>(x.t));
   }
+  void Unparse(const OmpToClause &x) {
+    auto &expect =
+        std::get<std::optional<std::list<OmpToClause::Expectation>>>(x.t);
+    auto &iter = std::get<std::optional<std::list<OmpIteratorModifier>>>(x.t);
```
Comment (kiranchandramohan) [inline flang/lib/Parser/unparse.cpp:2268]: Nit: braced initialisation

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 4]
```diff
@@ -268,6 +269,57 @@ bool OmpStructureChecker::IsCloselyNestedRegion(const OmpDirectiveSet &set) {
...
+  template <typename T> std::optional<bool> Visit(const T &) {
+    // Everything else.
+    return std::nullopt;
+  }
+
+private:
+  SemanticsContext &sctx_;
+  evaluate::FoldingContext &fctx_;
+};
+} // namespace
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:302]: Why do we need this? I see that there is `IsContiguous` that works in check-expression.

**Review by kiranchandramohan — COMMENTED** [review]
A few trivial comments.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 4 above)
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp:302]: It only works for certain classes in `evaluator`, specifically the ones in `variable.h`, i.e. the ones representing memory access.  This code extracts that from the top-level `Expr` class.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 5]
```diff
@@ -1647,24 +1700,29 @@ void OmpStructureChecker::CheckOrderedDependClause(
...
+        std::get<parser::OmpObjectList>(toClause.t), toSymbols);
+
     for (auto &[symbol, source] : toSymbols) {
-      auto fromSymbol = fromSymbols.find(symbol);
+      auto fromSymbol{fromSymbols.find(symbol)};
       if (fromSymbol != fromSymbols.end()) {
         context_.Say(source,
-            "A list item ('%s') can only appear in a TO or FROM clause, but not in both."_err_en_US,
+            "A list item ('%s') can only appear in a TO or FROM clause, "
+            "but not in both."_err_en_US,
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp]: Undone.

File: flang/lib/Parser/openmp-parsers.cpp (hunk 2 above)
Comment (kparzysz) [reply flang/lib/Parser/openmp-parsers.cpp]: Done.

File: flang/lib/Parser/unparse.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Parser/unparse.cpp]: Done.

File: flang/lib/Parser/unparse.cpp (hunk 3 above)
Comment (kparzysz) [reply flang/lib/Parser/unparse.cpp]: Done.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Comment by kazutakahirata** [comment]
I've fixed warnings from this PR with 76c16651c9f003332bff473d6703cf5bef367c99.  I didn't delete `sctx` in case you are planning to use it in the future.  Thanks!

**Pull request merged**